### PR TITLE
Added input action handler interface to receive action events regardless of the input source

### PR DESCRIPTION
**Pull request by luis-valverde-ms**
## Overview

Added a new input action handler interface to listen for input action events regardless of input source: 

```
public interface IMixedRealityInputActionHandler : IMixedRealityBaseInputHandler
{
	void OnActionStarted(BaseInputEventData eventData);
	void OnActionEnded(BaseInputEventData eventData);
}
```
Also added an input action handler script for convenience:

InputActionHandler

And an input actions example:

InputActionsExample

## Changes
- Fixes: #4006  .

### CODE REVIEW

**Comment by luis-valverde-ms** [comment]
@jbienzms , do you mind having a look at it and tell me what you think? I'll add the interfaces for actions that contain data soon but it should be easy enough to imagine how they will look like.

**Comment by julenka** [comment]
Talking to @killerantz Interactable isn’t actually meant to be used as the main way to listen to events. As I understand, the main power of interactable is in configuring visual states of buttons in response to events, not to hook up functionality in response to events.

Glad to see there is a way people may use instead of interactable.

**Comment by killerantz** [comment]
I think this is what I've been looking for all along. As long as I can get the source of the event, than this is what I was expecting from the Input Actions system. I wanted to build Interactable off of action events.

It would be great for Interaction Mappings to have a better link to the actions they invoke. or the eventData to include the InteractionMapping that invoked it as a property.

**Comment by jbienzms** [comment]
> @jbienzms , do you mind having a look at it and tell me what you think? I'll add the interfaces for actions that contain data soon but it should be easy enough to imagine how they will look like.

Soo sorry this has taken me so long. I'm in Berlin at hacks this week and didn't have a chance to look. Reviewing now.

File: Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs [hunk 1]
```diff
@@ -944,6 +944,25 @@ public void RaisePointerUp(IMixedRealityPointer pointer, MixedRealityInputAction
...
+                Debug.Assert(inputData.MixedRealityInputAction != MixedRealityInputAction.None);
+
+                var inputHandler = handler as IMixedRealityInputHandler;
+                if (inputHandler != null)
+                {
+                    inputHandler.OnInputDown(inputData);
+                }
+
+                var actionHandler = handler as IMixedRealityInputActionHandler;
+                if (actionHandler != null)
```
Comment (jbienzms) [inline Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: Do we also need to check if inputData.MixedRealityInputAction isn't **null**? I know we're checking to make sure it's not InputAction.None, but is there any point at which it could be **null**?

File: Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs [hunk 2]
```diff
@@ -967,6 +993,25 @@ public void RaiseOnInputDown(IMixedRealityInputSource source, Handedness handedn
...
+                Debug.Assert(inputData.MixedRealityInputAction != MixedRealityInputAction.None);
+
+                var inputHandler = handler as IMixedRealityInputHandler;
+                if (inputHandler != null)
+                {
+                    inputHandler.OnInputUp(inputData);
+                }
+
+                var actionHandler = handler as IMixedRealityInputActionHandler;
+                if (actionHandler != null)
```
Comment (jbienzms) [inline Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: Same question: Do we need to check for InputAction to be **null**?

File: Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs [hunk 3]
```diff
@@ -1103,12 +1155,39 @@ public void RaisePoseInputChanged(IMixedRealityInputSource source, Handedness ha
...
+                Debug.Assert(inputData.MixedRealityInputAction != MixedRealityInputAction.None);
+
+                var gestureHandler = handler as IMixedRealityGestureHandler;
+                if (gestureHandler != null)
+                {
+                    gestureHandler.OnGestureStarted(inputData);
+                }
+
+                var actionHandler = handler as IMixedRealityInputActionHandler;
+                if (actionHandler != null)
```
Comment (jbienzms) [inline Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: Same question: Do we need to check for InputAction to be **null**?

File: Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs [hunk 4]
```diff
@@ -1193,12 +1272,39 @@ public void RaiseGestureUpdated(IMixedRealityController controller, MixedReality
...
+                Debug.Assert(inputData.MixedRealityInputAction != MixedRealityInputAction.None);
+
+                var gestureHandler = handler as IMixedRealityGestureHandler;
+                if (gestureHandler != null)
+                {
+                    gestureHandler.OnGestureCompleted(inputData);
+                }
+
+                var actionHandler = handler as IMixedRealityInputActionHandler;
+                if (actionHandler != null)
```
Comment (jbienzms) [inline Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: Same question: Do we need to check for InputAction to be **null**?

File: Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs [hunk 5]
```diff
@@ -1287,6 +1393,26 @@ public void RaiseGestureCanceled(IMixedRealityController controller, MixedRealit
...
+                Debug.Assert(speechData.MixedRealityInputAction != MixedRealityInputAction.None);
+
+                var speechHandler = handler as IMixedRealitySpeechHandler;
+                if (speechHandler != null)
+                {
+                    speechHandler.OnSpeechKeywordRecognized(speechData);
+                }
+
+                var actionHandler = handler as IMixedRealityInputActionHandler;
+                if (actionHandler != null)
```
Comment (jbienzms) [inline Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: Same question: Do we need to check for InputAction to be **null**?

File: Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs [hunk 6]
```diff
@@ -1287,6 +1393,26 @@ public void RaiseGestureCanceled(IMixedRealityController controller, MixedRealit
...
+                {
+                    actionHandler.OnActionStarted(speechData);
+                    actionHandler.OnActionEnded(speechData);
```
Comment (jbienzms) [inline Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: Excellent. Yes, I agree. Speech should do Start and then End successively.

File: Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs (hunk 1 above)
Comment (luis-valverde-ms) [reply Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputSystem.cs]: `MixedRealityInputAction` is a value type so we're safe there.

File: Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/InputActionHandler.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        [SerializeField]
+        [Tooltip("Input Action to handle")]
+        private MixedRealityInputAction InputAction = MixedRealityInputAction.None;
+
+        [SerializeField]
+        [Tooltip("Whether input events should be marked as used after handling so other handlers in the same game object ignore them")]
+        private bool MarkEventsAsUsed = false;
+
+        [System.Serializable]
+        public class InputActionUnityEvent : UnityEvent<BaseInputEventData> { }
```
[the remark this answers is not in the document]
Comment (luis-valverde-ms) [reply Assets/MixedRealityToolkit.SDK/Features/Input/Handlers/InputActionHandler.cs]: Good idea. Changes coming.

**Comment by luis-valverde-ms** [comment]
> In this PR it looks like this could be accomplished in 2 ways:
> 
> Look at BaseInputEventData.Source
> Try cast the BaseInputEventData to the source-specific event data.
> 
> Am I correct about both of those working? If so, I think we're set in this regard.

@jbienzms , you're totally correct.

**Comment by luis-valverde-ms** [comment]
> It would be great for Interaction Mappings to have a better link to the actions they invoke. or the eventData to include the InteractionMapping that invoked it as a property.

@killerantz , agreed. Do you mind opening a feature request for that?

**Comment by luis-valverde-ms** [comment]
Checking this in with support only for digital input actions. Created #4576 to add support for non-digital ones, as there's a bit more work involved than expected.

File: Assets/MixedRealityToolkit/Interfaces/InputSystem/Handlers/IMixedRealityGestureHandler.cs [hunk 8]
```diff
@@ -8,7 +8,7 @@ namespace Microsoft.MixedReality.Toolkit.Input
     /// <summary>
     /// Interface to implement for generic gesture input.
     /// </summary>
-    public interface IMixedRealityGestureHandler : IEventSystemHandler
+    public interface IMixedRealityGestureHandler : IMixedRealityBaseInputHandler
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/Interfaces/InputSystem/Handlers/IMixedRealityGestureHandler.cs]: Why was this added?

Comment (luis-valverde-ms) [reply Assets/MixedRealityToolkit/Interfaces/InputSystem/Handlers/IMixedRealityGestureHandler.cs]: Have a look at the documentation of IMixedRealityBaseInputHandler and the implementation of OnGestureStartedWithAction() in MixedRealityInputSystem and let me know if that leaves you with any doubts.
